astropy/utils/tests/test_data.py

```diff
@@ -127,102 +429,265 @@ def test_download_cache():
+    urls = []
+    # tmpdir is shared between many tests, and that can cause weird
+    # interactions if we set the temporary cache too directly
+    with paths.set_temp_cache(tmpdir):
+        for i in range(5):
+            um, c = next(valid_urls)
+            assert not is_url_in_cache(um)
+            urls.append((um, c))
+        rs = download_files_in_parallel(
+            [u for (u, c) in urls], multiprocessing_start_method=method
+        )
+        assert len(rs) == len(urls)
+        url_set = set(u for (u, c) in urls)
+        assert url_set <= set(get_cached_urls())
+        for r, (u, c) in zip(rs, urls):
+            assert get_file_contents(r) == c
+        check_download_cache()
+    assert not url_set.intersection(get_cached_urls())
+    check_download_cache()
+
+
+def test_download_parallel_with_empty_sources(valid_urls, temp_cache):
+    urls = []
+    sources = {}
+    for i in range(5):
+        um, c = next(valid_urls)
+        assert not is_url_in_cache(um)
+        urls.append((um, c))
+    rs = download_files_in_parallel([u for (u, c) in urls], sources=sources)
+    assert len(rs) == len(urls)
+    # u = set(u for (u, c) in urls)
+    # assert u <= set(get_cached_urls())
+    check_download_cache()
+    for r, (u, c) in zip(rs, urls):
+        assert get_file_contents(r) == c
+
+
+def test_download_parallel_with_sources_and_bogus_original(
+    valid_urls, invalid_urls, temp_cache
+):
+    u, c = next(valid_urls)
+    urls = [(u, c, None)]
+    sources = {}
+    for i in range(5):
+        um, c_bad = next(valid_urls)
+        assert not is_url_in_cache(um)
+        sources[um] = []
+        for j in range(i):
+            sources[um].append(next(invalid_urls))
+        u, c = next(valid_urls)
+        sources[um].append(u)
+        urls.append((um, c, c_bad))
+    rs = download_files_in_parallel([u for (u, c, c_bad) in urls], sources=sources)
+    assert len(rs) == len(urls)
+    # u = set(u for (u, c, c_bad) in urls)
+    # assert u <= set(get_cached_urls())
+    for r, (u, c, c_bad) in zip(rs, urls):
+        assert get_file_contents(r) == c
+        assert get_file_contents(r) != c_bad
+
+
+def test_download_parallel_many(temp_cache, valid_urls):
+    td = []
+    for i in range(n_parallel_hammer):
+        u, c = next(valid_urls)
+        clear_download_cache(u)
+        td.append((u, c))
+
+    r = download_files_in_parallel([u for (u, c) in td])
+    assert len(r) == len(td)
+    for r, (u, c) in zip(r, td):
+        assert get_file_contents(r) == c
+
+
+def test_download_parallel_partial_success(temp_cache, valid_urls, invalid_urls):
+    td = []
+    for i in range(n_parallel_hammer):
+        u, c = next(valid_urls)
+        clear_download_cache(u)
+        td.append((u, c))
+
+    u_bad = next(invalid_urls)
+
+    with pytest.raises(urllib.request.URLError):
+        download_files_in_parallel([u_bad] + [u for (u, c) in td])
+    # Actually some files may get downloaded, others not.
```

Comment:
Four behaviours are reasonable(ish) when one URL is unavailable:

1. Nothing is added to the cache unless all succeed.
2. Every URL that works is added to the cache (and the user doesn't see the exception until they are all finished).
3. The URLs that are finished downloading are added to the cache, those "in flight" are aborted, and the user sees the exception right away. (This would require care to avoid leaving the cache locked, if it were possible at all; more likely possible with threads than multiprocessing, but this is IO bound anyway.)
4. When an error occurs no new downloads are started but any that are in progress are allowed to finish (or hang and timeout). The user sees the exception when all this is cleaned up (maybe some time later in case of a network blackout while downloads are in flight), and any downloads that finish successfully end up in the cache.

I think we currently get option 4. I don't know which we want, probably 3, but I don't know if we can do that (cancelling threads is hard, cancelling processes I don't know - could you just send them a SIGINT?).

Reply:
I can see people voting for each of those, LOL. We don't have to change the behavior.